### PR TITLE
[crypto] Add a C interface for SHA-512.

### PR DESCRIPTION
**Pull request by jadephilipoom**
The basic core here will work for SHA-384 as well, just with different initial values.

It's passing local tests but I'll merge those in a separate PR for diff-size reasons.

### CODE REVIEW

File: sw/device/lib/crypto/impl/sha2/sha512.c [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+status_t sha512_update(sha512_state_t *state, const uint8_t *msg,
+                       const size_t msg_len) {
+  // Process new data with no padding.
+  return process_message(state, msg, msg_len, kHardenedBoolFalse);
+}
+
+status_t sha512_final(sha512_state_t *state, uint8_t *digest) {
+  // Construct padding.
+  HARDENED_TRY(process_message(state, NULL, 0, kHardenedBoolTrue));
+  memcpy(digest, state->H, kSha512StateBytes);
```
Comment (moidx) [inline sw/device/lib/crypto/impl/sha2/sha512.c:324]: Are there any fields in `state` that need `hardened_memshred` and `memset -> 0`? Same for the one-shot function.

**Review by moidx — APPROVED** [review]

File: sw/device/lib/crypto/impl/sha2/sha512.c [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+static const otbn_app_t kOtbnAppSha512 = OTBN_APP_T_INIT(run_sha512);
+static const otbn_addr_t kOtbnVarSha512State =
+    OTBN_ADDR_T_INIT(run_sha512, state);
+static const otbn_addr_t kOtbnVarSha512Msg = OTBN_ADDR_T_INIT(run_sha512, msg);
+static const otbn_addr_t kOtbnVarSha512NChunks =
+    OTBN_ADDR_T_INIT(run_sha512, n_chunks);
+
+void sha512_init(sha512_state_t *state) {
+  // Set the initial state.
+  memcpy(state->H, kSha512InitialState, sizeof(state->H));
```
[the remark this answers is not in the document]
Comment (jadephilipoom) [reply sw/device/lib/crypto/impl/sha2/sha512.c]: Fixed. Right now the core OTBN operation isn't hardened, so I haven't put a lot of effort into hardening the Ibex side, but `hardened_memcpy` is a simple enough switch and I think it makes sense here.

File: sw/device/lib/crypto/impl/sha2/sha512.c [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+}
+
+/**
+ * Calculate an updated total message length and check that it is acceptable.
+ *
+ * Returns `OTCRYPTO_BAD_ARGS` if the current message length plus the new
+ * message length exceeds the SHA-512/SHA-384 maximum of 2^128 bits.
+ *
+ * This function is careful not to modify the state object; it should be
+ * modified as the very last step to avoid getting out of sync in case of
```
[the remark this answers is not in the document]
Comment (jadephilipoom) [reply sw/device/lib/crypto/impl/sha2/sha512.c]: Changed the wording here, hopefully it's more clear now.

File: sw/device/lib/crypto/impl/sha2/sha512.c (hunk 1 above)
Comment (jadephilipoom) [reply sw/device/lib/crypto/impl/sha2/sha512.c:341]: Good point, I added a `state_shred` function that gets called after the digest is copied.

**Pull request merged**